modules/ztp-nbde-rekeying-all-nbde-nodes.adoc

```diff
@@ -0,0 +1,124 @@
+          initialDelaySeconds: 30
+          periodSeconds: 10
+        env:
+        - name: ROOT_DEV
+          value: /dev/sda4
+        - name: NEW_TANG_PIN
+          value: >-
+            {"t":1,"pins":{"tang":[
+              {"url":"http://tangserver01:7500","thp":"WOjQYkyK7DxY_T5pMncMO5w0f6E"},
+              {"url":"http://tangserver02:7500","thp":"I5Ynh2JefoAO3tNH9TgI4obIaXI"},
+              {"url":"http://tangserver03:7500","thp":"38qWZVeDKzCPG9pHLqKzs6k1ons"}
+            ]}}
+        volumeMounts:
+        - name: hostroot
+          mountPath: /host
+        securityContext:
+          privileged: true
+      volumes:
+      - name: hostroot
+        hostPath:
+          path: /
+      nodeSelector:
+        kubernetes.io/os: linux
+      priorityClassName: system-node-critical
+      restartPolicy: Always
+      serviceAccount: machine-config-daemon
+      serviceAccountName: machine-config-daemon
+----
++
+In this case, even though you are rekeying `tangserver01`, you must specify not only the new thumbprint for `tangserver01`, but also the current thumbprints for all other Tang servers.  Failure to specify all thumbprints for a rekeying operation opens up the opportunity for a man-in-the-middle attack.
+
+. Distribute the daemonset to every cluster that needs to be rekeyed. This can be as simple as executing `oc apply -f tang-rekey.yaml` for a single cluster, but to execute at scale, wrap the daemonset in an ACM policy, with a placement rule to apply it to the appropriate set of clusters.
+
+[NOTE]
+====
+After validating that the daemonset has successfully rekeyed all servers, delete the daemonset. If you do not delete the daemonset, it must be deleted before the next rekeying operation.
+====
+
+.Validation
+
+After you distribute the daemonset, it is important to monitor the daemonsets to ensure that the rekeying has completed successfully. The script in the example daemonset terminates with an error if the rekeying failed, and remains in the `CURRENT` state if successful. There is also a readiness probe that marks the pod as `READY` when the rekeying has completed successfully.
+
+This is an example of the output listing for the daemonset before the rekeying has completed:
+
+[source,terminal]
+----
+$ oc get -n openshift-machine-config-operator ds rekey-tang
+NAME         DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
+rekey-tang   1         1         0       1            0           kubernetes.io/os=linux   11s
+----
+
+This is an example of the output listing for the daemonset after the rekeying has completed successfully:
+
+[source,terminal]
+----
+$ oc get -n openshift-machine-config-operator ds rekey-tang
+NAME         DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
+rekey-tang   1         1         1       1            1           kubernetes.io/os=linux   13h
+----
+
+Rekeying usually takes a few minutes to complete.
```

Comment:
After discussing this with the ACM people, I'd like to add a paragraph at the end:

> If you used ACM policies to distribute the daemonsets to multiple clusters, it should include a compliance policy that checks every daemonset’s `READY` count is equal to the `DESIRED` count. This way compliance to such a policy can demonstrate that all daemonset pods are `READY` and the re-keying has completed successfully. You could also use an ACM search to query all the daemonsets' states.

Reply:
Added this as a note.